clab/docker.go

```diff
@@ -179,6 +187,52 @@ func (c *cLab) CreateContainer(ctx context.Context, node *Node) (err error) {
 	return createContainerNS(node.Pid, node.LongName)
 }
 
+func (c *cLab) PullImageIfRequired(ctx context.Context, node *Node) (err error) {
+	filter := filters.NewArgs()
+	filter.Add("reference", node.Image)
+
+	ilo := types.ImageListOptions{
+		All:     false,
+		Filters: filter,
+	}
+
+	log.Debugf("Looking up %s Docker image", node.Image)
+
+	images, err := c.DockerClient.ImageList(ctx, ilo)
+	if err != nil {
+		return err
+	}
+
+	// Image doesn't exist, need to pull it
+	if len(images) == 0 {
+		// might need canonical name e.g.
+		//    -> alpine == docker.io/library/alpine
+		//    -> foo/bar == docker.io/foo/bar
+		//    -> docker.elastic.co/elasticsearch/elasticsearch == docker.elastic.co/elasticsearch/elasticsearch
+		canonicalImageName := node.Image
+		slashCount := strings.Count(node.Image, "/")
+
+		switch slashCount {
+		case 0:
+			canonicalImageName = "docker.io/library/" + node.Image
+		case 1:
+			canonicalImageName = "docker.io/" + node.Image
+		}
+
+		log.Infof("Pulling %s Docker image", node.Image)
+		reader, err := c.DockerClient.ImagePull(ctx, canonicalImageName, types.ImagePullOptions{})
```

Comment:
add `defer reader.Close()` after you created a readcloser on L223

Reply:
Thanks, will do!